File: src/cfiles/MDHLink.cpp
#include "stdafx.h"
#include "MDHLink.h"


MDHLink::MDHLink(double theta, double d, double a, double alpha, int type, double offset):
	_theta(theta),_d(d), _a(a), _alpha(alpha), _type(type),_offset(offset)
{
}


void MDHLink::Transform(double q)
{
	if (_type==e_rotation)
		_theta = q+_offset;
	else
		_d = q+_offset;

	double st = sin(_theta); double sa = sin(_alpha);
	double ct = cos(_theta); double ca = cos(_alpha);
	_pose.rot(0, 0) = ct;
	_pose.rot(0, 1) = -st;
	_pose.rot(0, 2) = 0;

	_pose.rot(1, 0) = st*ca;
	_pose.rot(1, 1) = ct*ca;
	_pose.rot(1, 2) = -sa;

	_pose.rot(2, 0) = st*sa;
	_pose.rot(2, 1) = ct*sa;
	_pose.rot(2, 2) = ca;

	_pose.pos<<_a, -sa*_d, ca*_d;
}

MDHLink::~MDHLink()
{
}

File: src/classes/MDHLink.m
classdef MDHLink < handle
    
    properties
        theta
        d
        a
        alpha
        type
        offset
    
        m
        cog
        inertia
        fv
        fc
        
        pose
    end
    
    methods
        function obj = MDHLink(dh)
            obj.theta = dh(1);
            obj.d = dh(2);
            obj.a = dh(3);
            obj.alpha = dh(4);
            obj.type = dh(5);
            obj.offset = dh(6);

            obj.m = 0;
            obj.cog = zeros(3,1);
            obj.inertia = zeros(3,3);
            obj.fv = 0;
            obj.fc = 0;

            obj.pose = eye(4);            
        end
    
        function Transform(obj, q)
            if obj.type==0
                obj.theta = q+obj.offset;
            else
                obj.d = q+obj.offset;
            end
            st = sin(obj.theta);
            sa = sin(obj.alpha);
            ct = cos(obj.theta);
            ca = cos(obj.alpha);
            rot(1,1) = ct; rot(1,2) = -st; rot(1,3) = 0;
            rot(2,1) = st*ca; rot(2,2) = ct*ca; rot(2,3) = -sa;
            rot(3,1) = st*sa; rot(3,2) = ct*sa; rot(3,3) = ca;
            pos(1) = obj.a; pos(2) = -sa*obj.d; pos(3) = ca*obj.d;
            obj.pose(1:3,1:3) = rot;
            obj.pose(1:3,end) = pos;
        end

    end
    
    
    
end
